fabric-biomes-v1/src/main/java/net/fabricmc/fabric/mixin/biomes/MixinAddHillsLayer.java

```diff
@@ -0,0 +1,75 @@
+/*
+ * Copyright (c) 2016, 2017, 2018, 2019 FabricMC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package net.fabricmc.fabric.mixin.biomes;
+
+import net.fabricmc.fabric.api.biomes.v1.OverworldBiomes;
+import net.fabricmc.fabric.impl.biomes.InternalBiomeData;
+import net.fabricmc.fabric.impl.biomes.WeightedBiomePicker;
+import net.minecraft.util.registry.Registry;
+import net.minecraft.world.biome.Biome;
+import net.minecraft.world.biome.layer.AddHillsLayer;
+import net.minecraft.world.biome.layer.BiomeLayers;
+import net.minecraft.world.biome.layer.LayerRandomnessSource;
+import net.minecraft.world.biome.layer.LayerSampler;
+import org.spongepowered.asm.mixin.Mixin;
+import org.spongepowered.asm.mixin.injection.At;
+import org.spongepowered.asm.mixin.injection.Inject;
+import org.spongepowered.asm.mixin.injection.callback.CallbackInfoReturnable;
+
+/**
+ * Injects hills biomes specified from {@link OverworldBiomes#addHillsBiome(Biome, Biome, int)}into the default hills layer
+ */
+@Mixin(AddHillsLayer.class)
+public class MixinAddHillsLayer {
+
+	@Inject(at = @At("HEAD"), method = "sample", cancellable = true)
+	private void sample(LayerRandomnessSource rand, LayerSampler biomeSampler, LayerSampler noiseSampler, int chunkX, int chunkZ, CallbackInfoReturnable<Integer> info) {
+		final int biomeId = biomeSampler.sample(chunkX, chunkZ);
+		int noiseSample = noiseSampler.sample(chunkX, chunkZ);
+		int processedNoiseSample = (noiseSample - 2) % 29;
+		final Biome biome = Registry.BIOME.get(biomeId);
+
+		if (InternalBiomeData.getOverworldHills().containsKey(biome) && (rand.nextInt(3) == 0 || processedNoiseSample == 0)) {
```

Comment:
Player has explained the issue on Discord, this condition can evaluate to `false` after calling `rand.nextInt()`. If that's the case, vanilla logic will be executed, but the LayerRandomnessSource object will not be in the same state. The goal of letting vanilla running after this injection is to have exactly the same results as if the injection was not here at all, but touching the random means future generated numbers will be different, making the results different. 
As such, something needs to be done to make sure the randomness source is in the same state at the beginning and at the end of the injection when no modded biome is selected.

Reply:
Ah, I believe that was a change prospector made

I can change the order of stuff probably

Reply:
wait a second

Reply:
I see what you mean

Reply:
Maybe it should be changed into a return mixin

Reply:
I'm tired at the moment so I probably won't think of a good solution